Expose climate control to home assistant

iamer opened the pull request:
This pull request implements climate control as discussed in #54

The climate entity exposed to home assistant worked when I tested it, though probably still has quirky behaviour.
 
I used Claude Sonnet 4.5 LLM for the initial implementation because I am not familiar with home assistant integration code and APIs.

iamer committed:
Implement climate control

* Coded with support from Claude Sonnet 4.5
* Tested with Toyota CHR 2025

iamer committed:
Use vehicle property climate_settings

iamer committed:
Address automated code review comments

iamer committed:
Refactor _create_climate_settings to appease Codacy

iamer committed:
Add current temperature and state sync

iamer commented:
In my testing the temperature control works, and at the least the front defrost also works. I suspect that the rear defrost is not working reliably. Both defrost options are available in the official app and work for me, so it could be unsupported for your car model?

Testing changes is slow for multiple reasons:
1. If I try to heat two times in a row the server refuses the second heating command with an error that the car needs to be started manually first.
2. I don't want to waste fuel and stress the engine by restarting it multiple times in a row.
3. Unlocking the car during a heating cycle  to check the climate settings turns off the engine and heating. This is an annoying behaviour that I have complained to the Toyota dealer about and they "acknowledged" the complaint.

I think adding a lot of debugging log messages could help figuring out what is happening and speed up the implementation process.

werner-j commented:
Hi. I got your additional stuff and it works flawlessly for me. Over the vacation I want to give it a try and see how It behaves. I noticed that it's not "resetting" the climate to "off" when I open the car or the 20 minutes are over that Toyota sets for heating or cooling. Is it possible to get a callback from the car, that the heating turned off and show that in Home Assistant? I can help myself for the time being with an automation, but of course it would be nicer if the real state would be reflected in Home Assistant.

Thanks for the great work!

iamer commented:
Thanks for the nice feedback, and great that it works for you.

Yes, I mentioned above that the climate device will go out of sync if the climate sate changes externally (car started, or after the default max of 20 minutes). The underlying pytoyoda python module does not provide a callback mechanism for the climate status and I am not yet sure if it is possible to implement it that way.

However, I am testing an alternate approach based on polling the climate status from the car when the climate device is on until it is turned off.  I will push the code once I am satisfied with how it works.

Paja-git commented:
Hi @iamer , I just started to test your climate PR. For the Lexus brand (NX 450h+). 
e simple copy of the climate.py didn't work , the Climate control was not present. I had to change the Toyota/const.py line to import the Platform.CLIMATE to supported ones. After the HA reboot, the Climate control is there. 
The change is: 
```
PLATFORMS = [Platform.BINARY_SENSOR, Platform.DEVICE_TRACKER, Platform.SENSOR, Platform.CLIMATE]
```
Maybe it is because the Lexus control is a very first Climate platform control in my HA instance?

iamer committed:
Attempt to fix defrost settings and add climate to supported platforms

iamer commented:
I also added an attempt at fixing the defrost settings.

iamer committed:
Scale back logging to debug level

iamer committed:
Refactor and simplify the code

ahorseman commented:
> I have pushed the refactored code after testing it. It is now much simpler. Please give it a try.

I am on vacation until 10th of January. After I'm back I'll give it another try. Thank you so much for your effort put into this. Really great work!

iamer commented:
Hi, please make sure you have the latest changes always from here: https://github.com/pytoyoda/ha_toyota/pull/221/changes

What other sensors / actions do you mean?

adriannelson commented:
Thankyou @iamer! Easy when you know how, works straight away :100:

ArnstadFredrik commented:
I have used the climate controls for my BZ4X for about a week and it works.

A couple of points of general feedback.
- The climate control state (temp and options) reset to (21°C, none) after home assistant restart.
- Other clate controls like seat and steering wheel heating is bot an option. That in itself is not a huge problem, but the way the car handles starting the cars climate without them specified, will turn them `off` not leve them on `auto`. Many more options in the dropdown could become cluttered. Optionaly seat and steering wheel could just be included with the current defrost options.

hanseartic commented:
> > Really nice addition 🙏🏼.
> > I am trying this out with our EV. There are two remarks, already:
> > 
> > * our car also has seat-heating (driver, passenger can be chosen individually) and steering wheel heating options in the app
> 
> Strange, my EV also has seat and steering wheel heating, same as you describe. But I don't get that as options from the api endpoint as you do.

I gathered the response with the [`simple_client_example.py`](https://github.com/pytoyoda/pytoyoda/blob/e7471ea49e3b87f657d1f5afac3b0addfb4ccc87/sandbox/simple_client_example.py#L70). Uncommented the `get_climate_settings`.

iamer committed:
Implement settings preserving and persistence

iamer committed:
refactor AC operations handling

iamer committed:
fix defrost parameters handling

iamer committed:
implement debounce and delayed settings sync to car

iamer committed:
use home assistant callback and async functions

iamer committed:
Fix codacy error

Method _load_climate_settings_from_coordinator has a cyclomatic complexity of 12 (limit is 8)

iamer committed:
Add translations

iamer committed:
Skip climate entity creation for vehicles without climate capability

Only create climate entities for vehicles that advertise the
climate_start_engine capability in their features. This prevents
creating non-functional climate entities for vehicles that don't
support remote climate control.

MaBeniu commented:
This working great. Confirm after testing on lexus ux. Great job!! Thanks so much!

This integration for my use misses only lock control. Otherwise finally getting to very nice usable integration.

thePier commented:
Love the climate functionality finally exposed!

Just one question: I added the climate entity onto my "car" dashboard and it shows as just a button like this
<img width="319" height="282" alt="image" src="https://github.com/user-attachments/assets/79c388fa-7afb-4096-95ac-c300fd864290" />

I was wondering how can I make it show like HA shows the actual climate controls (the -/+ and the set temperature) in its default Areas dashobard
<img width="174" height="109" alt="image" src="https://github.com/user-attachments/assets/838eca8c-b87a-4ca4-a694-1accf7b867e3" />

iamer commented:
You can try using a thermostat card for example?

thePier commented:
it works with the Thermostat card, I expected for the entity to be already configured as a Thermostat.
Thanks!
